File: src/services/tags.ts
import type { Env, Viewer, DbMemo } from "../types";
import { json, readJson, safeJsonParse, unixNow } from "../utils";
import { buildMemoPayload } from "./memo";
import { recordAudit } from "./audit";

export function normalizeTagName(value: string): string {
  return value.trim().replace(/^#/, "").replace(/\s+/g, "-").slice(0, 64);
}

export function replaceTagInContent(content: string, from: string, to: string): string {
  const escaped = from.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return content.replace(new RegExp(`(^|\\s)#${escaped}(?=$|\\s)`, "gu"), `$1#${to}`);
}

export async function listTags(env: Env, viewer: Viewer): Promise<Response> {
  const where = viewer.role === "ADMIN" ? "" : "WHERE memo.creator_id = ?";
  const rows = await env.DB.prepare(`
    SELECT memo.id, memo.payload
    FROM memo
    ${where}
  `).bind(...(viewer.role === "ADMIN" ? [] : [viewer.id])).all<{ id: number; payload: string }>();

  const counts = new Map<string, number>();
  for (const row of rows.results) {
    const payload = safeJsonParse(row.payload, {}) as { tags?: string[] };
    for (const tag of payload.tags ?? []) counts.set(tag, (counts.get(tag) ?? 0) + 1);
  }

  return json({
    tags: [...counts.entries()]
      .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))
      .map(([name, count]) => ({ name, count }))
  });
}

export async function renameTag(request: Request, env: Env, viewer: Viewer): Promise<Response> {
  const body = await readJson<{ from?: string; to?: string }>(request);
  const from = normalizeTagName(body.from ?? "");
  const to = normalizeTagName(body.to ?? "");
  if (!from || !to) return json({ error: "from and to are required" }, 400);

  const rows = await env.DB.prepare(`
    SELECT * FROM memo WHERE content LIKE ? ${viewer.role === "ADMIN" ? "" : "AND creator_id = ?"}
  `).bind(`%#${from}%`, ...(viewer.role === "ADMIN" ? [] : [viewer.id])).all<DbMemo>();

  let updated = 0;
  for (const memo of rows.results) {
    const nextContent = replaceTagInContent(memo.content, from, to);
    if (nextContent === memo.content) continue;
    await env.DB.prepare("UPDATE memo SET content = ?, payload = ?, updated_ts = ? WHERE id = ?")
      .bind(nextContent, JSON.stringify(buildMemoPayload(nextContent)), unixNow(), memo.id)
      .run();
    updated += 1;
  }
  await recordAudit(env, viewer, "tag.rename", from, { to, updated });
  return json({ updated });
}

File: src/tags.rs
use super::*;

pub(crate) async fn list_tags(
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    let memos = get_recent_memos(env, viewer, 500).await?;
    let mut counts: HashMap<String, i64> = HashMap::new();
    for memo in memos {
        if let Ok(payload) = serde_json::from_str::<Value>(&memo.payload) {
            if let Some(tags) = payload.get("tags").and_then(Value::as_array) {
                for tag in tags.iter().filter_map(Value::as_str) {
                    *counts.entry(tag.to_string()).or_default() += 1;
                }
            }
        }
    }
    let tags: Vec<Value> = counts
        .into_iter()
        .map(|(name, count)| json!({ "name": name, "count": count }))
        .collect();
    json_response(json!({ "tags": tags }), 200).map_err(AppError::from)
}

pub(crate) async fn rename_tag(
    req: &mut Request,
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    let body: Value = req
        .json()
        .await
        .map_err(|_| AppError::new(400, "Invalid JSON"))?;
    let from = normalize_tag_name(body.get("from").and_then(Value::as_str).unwrap_or(""));
    let to = normalize_tag_name(body.get("to").and_then(Value::as_str).unwrap_or(""));
    if from.is_empty() || to.is_empty() {
        return Err(AppError::new(400, "from and to are required"));
    }
    let rows = if viewer.role == "ADMIN" {
        db(env)?.prepare("SELECT memo.*, \"user\".username AS creator_username, \"user\".nickname AS creator_nickname FROM memo JOIN \"user\" ON \"user\".id = memo.creator_id WHERE memo.content LIKE ?")
            .bind(&[format!("%#{}%", from).into()])?
            .all()
            .await?
    } else {
        db(env)?.prepare("SELECT memo.*, \"user\".username AS creator_username, \"user\".nickname AS creator_nickname FROM memo JOIN \"user\" ON \"user\".id = memo.creator_id WHERE memo.content LIKE ? AND memo.creator_id = ?")
            .bind(&[format!("%#{}%", from).into(), js_num(viewer.id)])?
            .all()
            .await?
    };
    let memos: Vec<DbMemo> = rows.results()?;
    let mut updated = 0;
    for memo in memos {
        let next_content = replace_tag_in_content(&memo.content, &from, &to);
        if next_content == memo.content {
            continue;
        }
        db(env)?
            .prepare("UPDATE memo SET content = ?, payload = ?, updated_ts = ? WHERE id = ?")
            .bind(&[
                next_content.clone().into(),
                build_memo_payload(&next_content).to_string().into(),
                js_num(unix_now()),
                js_num(memo.id),
            ])?
            .run()
            .await?;
        updated += 1;
    }
    json_response(json!({ "updated": updated }), 200).map_err(AppError::from)
}

pub(crate) async fn timeline(
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    let rows = if viewer.role == "ADMIN" {
        db(env)?.prepare("SELECT date(created_ts, 'unixepoch') AS day, COUNT(*) AS count FROM memo WHERE row_status = 'NORMAL' GROUP BY day ORDER BY day DESC LIMIT 120")
            .all().await?
    } else {
        db(env)?.prepare("SELECT date(created_ts, 'unixepoch') AS day, COUNT(*) AS count FROM memo WHERE row_status = 'NORMAL' AND (visibility != 'PRIVATE' OR creator_id = ?) GROUP BY day ORDER BY day DESC LIMIT 120")
            .bind(&[js_num(viewer.id)])?
            .all().await?
    };
    let days: Vec<Value> = rows.results()?;
    json_response(json!({ "days": days }), 200).map_err(AppError::from)
}
